compiler/src/iree/compiler/Dialect/Stream/Transforms/SpecializeEncodings.cpp

```diff
@@ -82,6 +84,273 @@ getEncodingWithNewLayouts(Type type,
+      result.push_back(emptyArray);
+      continue;
+    }
+    result.push_back(b.getArrayAttr(llvm::to_vector_of<Attribute>(affinities)));
+  }
+  return result;
+}
+
+/// Returns the encoding layout of each binding. They are resolved by the
+/// `resolveLayoutAttr` and the corresponding affinity. An empty array attribute
+/// indicates that the operand resource does not have an encoding in the tensor
+/// type.
+static FailureOr<SmallVector<Attribute>> getEncodingLayoutForBindings(
+    ModuleOp moduleOp, FunctionOpInterface funcOp,
+    ArrayRef<Attribute> operandAffinities,
+    IREE::Stream::AffinityAttr resultAffinity,
+    IREE::Stream::ResolveLayoutAttrFn resolveLayoutAttr) {
+  MLIRContext *ctx = funcOp.getContext();
+  Region &region = funcOp.getFunctionBody();
+
+  // The size of function arguments could be greater than the number of operand
+  // affinities because the function also captures output resources in the
+  // arguments.
+  SmallVector<ArrayAttr> argsAffinities = llvm::map_to_vector(
+      operandAffinities, [](Attribute attr) { return cast<ArrayAttr>(attr); });
+  auto resAffinityAttr = ArrayAttr::get(ctx, {cast<Attribute>(resultAffinity)});
+  argsAffinities.resize(region.getNumArguments(), resAffinityAttr);
+
+  SmallVector<Attribute> result;
+  auto emptyArrayAttr = ArrayAttr::get(ctx, {});
+  int idx = 0;
+  for (auto arg : region.getArguments()) {
+    if (!isa<IREE::Stream::BindingType>(arg.getType())) {
+      continue;
+    }
+    ArrayRef<Attribute> affinities = argsAffinities[idx++].getValue();
+    assert(affinities.size() == 1);
+
+    SetVector<Attribute> layoutResolvers;
+    if (failed(
+            resolveLayoutAttr(cast<IREE::Stream::AffinityAttr>(affinities[0]),
+                              moduleOp, layoutResolvers))) {
+      return failure();
+    }
+
+    for (auto user : arg.getUsers()) {
+      auto subspanOp = dyn_cast<IREE::Stream::BindingSubspanOp>(user);
+      if (!subspanOp) {
+        return failure();
+      }
+      auto resultType = llvm::dyn_cast<IREE::Flow::DispatchTensorType>(
+          subspanOp.getResult().getType());
+      if (!resultType) {
+        return failure();
+      }
+      std::optional<IREE::Encoding::EncodingAttr> newEncodingType =
+          getEncodingWithNewLayouts(resultType.asRankedTensorType(),
+                                    layoutResolvers);
+      if (!newEncodingType) {
+        result.push_back(emptyArrayAttr);
+      } else {
+        result.push_back(newEncodingType.value());
+      }
+    }
+  }
+
+  return result;
+}
+
+/// Returns the resolved layouts of the bindings for the `dispatchOp`. It
+/// gathers the operand resource affinities and the execution affinity, looks at
+/// the bindings in executable functions, and resolves the layouts for the
+/// tensor types. Returns a failure if it can not figure the layouts.
+static FailureOr<SmallVector<Attribute>> getBindingLayoutsForDispatchOp(
```

Comment:
```suggestion
/// TODO(#19806): remove this walk of the executable and instead use the encodings directly on the `stream.tensor.dispatch` op.
///
/// Returns the resolved layouts of the bindings for the `dispatchOp`. It
/// gathers the operand resource affinities and the execution affinity, looks at
/// the bindings in executable functions, and resolves the layouts for the
/// tensor types. Returns a failure if it can not figure the layouts.
static FailureOr<SmallVector<Attribute>> getBindingLayoutsForDispatchOp(
```

reaching into the executable isn't great and something we need to fix - there is no `stream.tensor.dispatch` today because there previously hadn't been a need, but I'm thinking this is the need we were looking for. A `stream.tensor.dispatch` that also captured tensor encodings would let you tell from the dispatch sites and avoid the need to reach inside, and then in the encode tensors pass we'd swap it out with a `stream.async.dispatch`. I filed #19806 and you can tag it here as a TODO to remove this function (we'd just have `TensorDispatchOp::getBindingLayouts()` and `setBindingLayouts()` methods).

Reply:
implemented here: #19817 - should let you remove this and instead get all operand and result encodings (`getOperandEncodings() + getOperandEncodingDims() + getResultEncodings() + getResultEncodingDims()`)

Reply:
Cool! I think then I can also implement `updateOperandEncodings` + `updateResultEncoding` methods with my type interface prototype. It should make code cleaner. Thanks!

Reply:
Removed the TODO because now we can update the encodings in "preprocessing" and use them directly in the duplication.